nx/lib/nx/binary_backend.ex

```diff
@@ -2153,6 +2153,72 @@ defmodule Nx.BinaryBackend do
     from_binary(output, new_data)
   end
 
+  @impl true
+  def fft(out, %{shape: {n}} = tensor) do
+    data =
+      match_types [tensor.type] do
+        for <<match!(x, 0) <- to_binary(tensor)>> do
+          read!(x, 0)
+        end
+      end
+
+    result =
+      if rem(n, 2) == 0 do
+        fft_list(data)
+      else
+        dft_list(data)
+      end
```

Comment:
I think we can just use `dft_list` for all tensors. I'll try to compare both using benchee, but I don't expect `fft_list` to actually be faster. I'll bring back the results here

Reply:
> We can also add a pad option which pads the input tensor with zeros to the next power of two, which is a common DSP technique for best use of the FFT. WDYT?

Sounds good to me!

Reply:
Both [torch.fft.fft](https://pytorch.org/docs/stable/generated/torch.fft.fft.html#torch.fft.fft) and [numpy.fft.fft](https://numpy.org/doc/stable/reference/generated/numpy.fft.fft.html) have an `n` option that either trims or pads to the given size.

We could try implementing a more complex algorithm to optimise non-powers-of-2, but for performance people should be using a different backend anyway, so it's probably an overkill. Also, [a random reference](https://math.stackexchange.com/a/77152).

Reply:
It's easy to get the next power of 2 of a number through log and ceil, so maybe we can add a `pad: integer() | :power_of_two` option?

Reply:
Turns out Enum.zip was bugging out and the algorithm really only does work for powers of two. I've refactored to only apply the recursive version in said cases, and use the naive DFT otherwise.

I'll just add the padding option and this is ready for review again :)

Reply:
Oh interesting, why does it make a difference if we compute the recursive call with explicit DFT or FFT?

Reply:
The FFT algorithm removes a lot of the complex exponential calculations by some clever recombining of the parts which were recursively calculated. https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm

Since we have the Radix-2 implementation, it only works for powers of two, but there are versions which work in general length, but depend on prime factorizations and/or fast matrix multiplication, so I think it's probably overkill as well.